src/db/models.rs

```diff
@@ -81,41 +82,39 @@ impl User {
         bcrypt::hash(&password.into()).expect("Couldn't hash password!")
     }
 
-    // TODO: should probably be Result<Option<T>>?
-    // requires a bit of thought, because we can't just try_opt! then :(
-    pub fn by_username(db_conn: &Connection, username: String) -> Option<User> {
-        let account = try_opt!({
+    pub fn by_username(db_conn: &Connection, username: String) -> QueryResult<Option<User>> {
+        let account = try_resopt!({
             use db::schema::accounts::dsl;
             dsl::accounts
                 .filter(dsl::username.eq(username))
                 .filter(dsl::domain.is_null())
                 .first::<Account>(&**db_conn)
                 .optional()
-                .unwrap()
         });
 
         use db::schema::users::dsl;
-        dsl::users
+        Ok(dsl::users
             .filter(dsl::account_id.eq(account.id))
             .first::<User>(&**db_conn)
-            .optional()
-            .unwrap()
+            .optional()?)
     }
 }
 
 impl Account {
     // TODO: result
-    pub fn fetch_local_by_username<S>(db_conn: &Connection, username: S) -> Option<Account>
+    pub fn fetch_local_by_username<S>(
+        db_conn: &Connection,
+        username: S,
+    ) -> QueryResult<Option<Account>>
     where
         S: Into<String>,
     {
         use db::schema::accounts::dsl;
-        dsl::accounts
+        Ok(dsl::accounts
             .filter(dsl::username.eq(username.into()))
             .filter(dsl::domain.is_null())
             .first::<Account>(&**db_conn)
-            .optional()
-            .unwrap()
+            .optional()?)
```

Comment:
and here

Reply:
Thanks for pointing this out! I'd left these in from when I was testing stuff with `Carrier` and the `Fail` trait, back when these functions directly returned `failure::Error`.